### PR TITLE
Update/ballmer groupings 2024

### PR DESCRIPTION
**Pull request by coryamanda**
# Description

This PR onboards the custom aggregations we receive every year from the College Board as well as all the calculations that derive from those. It also moves the logic for generating the CDO 2+ race URG multiplier into Hydrone (find the previous version in _seed_files.yml in the description for "seed_ap_tr_urg_multiplier")

List of files:
- data/seed_ap_tr_urg_multiplier.csv - adds a hard copy of the 2024 multiplier so it doesn't change from year to year
- macros/ap_normalization_macros.sql - adds to the list of category names referenced in the AP macros
- models/reporting_views/ballmer/csa_csp_csd_started.sql - creates a version of csa_csp_csd_started to use in the multiplier calculation. The seed file contains multipliers for both started and completed, but right now only started is being used. 
- models/reporting_views/ballmer/2+race_urg_multiplier - moves the logic for calculating the multiplier into Hydrone. The previous version based in Redshift is linked for reference.
- models/staging/external_datasets/base/base_external_datasets__ap_agg_exam_results_2024.sql - ingest 2024 data
- models/staging/external_datasets/stg_external_datasets__ap_agg_exam_results.sql - adds 2024 data to the staging file. I know that best practice is for staging files to only reference a single table. I'm happy to fix that another time. 

## Links

Jira ticket(s): []()

## Testing story

Supporting document here: [<google-sheets-html-origin><style type="text/css"></style>](https://docs.google.com/spreadsheets/d/1hchifUCVhEqeR7kYqN6tE2EkR5Gd4gonmqkJAk1Uje4/edit?gid=1379317685#gid=1379317685)

- Was the data onboarded correctly?
Yes
- Did the manual changes to the macros successfully create demographic_category labels for the 2024 data?
Yes
- Are the calculations (bhnapi, two+ race as URG) working as correctly assuming that the multiplier is correct?
Yes
- Did the small clarity edits to the demographic categories cause any issues in the historical data?
No - when ordered, it's all the same except in the one situation where two categories are sorted differently alphabetically
- Can I recreate the Code.org multipliers that were generated last year using the Hydrone version rather than the Redshift one? 
Close - the CSA/CSP/CSD started numbers match with about a 1% drop from last year, which is about what we would expect due to the loss of students whose accounts are being deleted. The CSA/CSP completed numbers don't match the older versions, which make since I'm using CSA/CSP completed and the Redshift version is CSA/CSP/CSD completed

## Privacy

- [ ] 1.	Does this change involve the collection, use, or sharing of new Personal Data? No
- [ ] 2.    Do these data exist in the appropriate schema(s)? Yes
- [ ] 3.	Does this change involve a new or changed use or sharing of existing Personal Data? No
- [ ] 4.    Consider: will this data be visible on Tableau? will this data be surfaced in a report exported from Trevor? Eventually yes, in a new view supporting the existing AP dashboard
- [ ] 5.    If yes to any of the above, please list the models, columns, and justification below:
      i.
      ii.
      iii. 

## PR Checklist:
--> **Note: if these are not all checked, the PR will be sent back.**

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code adheres to [style guide](https://docs.getdbt.com/best-practices/how-we-style/0-how-we-style-our-dbt-projects)👀 and is **[DRY](https://docs.getdbt.com/terms/dry)**
- [ ] Code is well-commented (**please do not leave extraneous commentary in model code, if it is for the purpose of documentation, please relocate accordingly)
- [ ] Appropriate documentation has been provided (see `.yml.`, did `dbt docs generate` succeed?)
- [ ] New features are translatable or updates will not break up/downstream models
- [ ] Relevant documentation has been added or updated (i.e. `dbt docs` has been updated successfully on [Github Pages](code-dot-org.github.io/analytics/)
- [ ] Pull Request is labeled appropriately (eg. `chore/`, `feature/`, `fix/`)
- [ ] Follow-up work items (including potential tech debt) are tracked and linked (if applicable)

### CODE REVIEW

**Comment by coryamanda** [comment]
@nataliazm99 the testing story link didn't come through well in the PR. You can find the full testing story here: https://docs.google.com/spreadsheets/d/1hchifUCVhEqeR7kYqN6tE2EkR5Gd4gonmqkJAk1Uje4/edit?gid=1379317685#gid=1379317685

File: dbt/data/seed_ap_tr_urg_multiplier.csv [hunk 1]
```diff
@@ -6,10 +6,12 @@ ap_urg_calc_started,2020,2019-20,131551,296682,16624,23257,1.6120513869786437,us
...
 ap_urg_calc_completed,2017,2016-17,4418,12430,531,885,1.6880941602535084,analysis.csp_csd_completed per year,2024-05-15
 ap_urg_calc_completed,2018,2017-18,10446,28885,1361,2093,1.7980892610041277,analysis.csp_csd_completed per year,2024-05-15
 ap_urg_calc_completed,2019,2018-19,15856,44487,2135,3301,1.8146456742918928,analysis.csp_csd_completed per year,2024-05-15
 ap_urg_calc_completed,2020,2019-20,15143,43875,1955,3162,1.791389825598115,analysis.csp_csd_completed per year,2024-05-15
 ap_urg_calc_completed,2021,2020-21,16400,48929,1746,3084,1.6890883315934326,analysis.csp_csd_completed per year,2024-05-15
 ap_urg_calc_completed,2022,2021-22,18293,48837,1965,3329,1.5758424781275957,analysis.csp_csd_completed per year,2024-05-15
-ap_urg_calc_completed,2023,2022-23,19648,52648,2066,3590,1.5420533421647171,analysis.csp_csd_completed per year,2024-05-15
\ No newline at end of file
+ap_urg_calc_completed,2023,2022-23,19648,52648,2066,3590,1.5420533421647171,analysis.csp_csd_completed per year,2024-05-15
+ap_urg_calc_completed,2024,2023-24,4349,11764,494,858,1.55741828494185,using hydrone csa_csp_cmpltd per year,2024-12-10
```
Comment (nataliazm99) [inline dbt/data/seed_ap_tr_urg_multiplier.csv:17]:
```suggestion
ap_urg_calc_completed,2024,2023-24,4349,11764,494,858,1.55741828494185,using hydrone csa_csp_cmpleted per year,2024-12-10
```

File: dbt/models/reporting_views/ballmer/csa_csp_csd_started.sql
Comment (nataliazm99) [inline dbt/models/reporting_views/ballmer/csa_csp_csd_started.sql:1]: Use this instead, pulling from `dim_user_course_activity` which has all the aggregations already. This ensures consistency and improves performance and cross-checking.

``` 
with duca as (
  select * 
    from 
    {{ref('dim_user_course_activity')}}
      where user_type = 'student'
    and course_name in ('csa','csp','csd')
    ),
    

    final as (
        select 
        course_name
        , user_id
        , school_year
        , first_activity_at	as started_at
        , last_activity_at	as last_progress_at
        , num_levels		as lvl_cnt
    from duca 
        )

    select * from final
```

**Review by nataliazm99 — COMMENTED** [review]
Looks good to me, the only changes I would make are suggested in the comments. Additionally, would it help to have the school_id in `ballmer_aggregates_2024` to avoid joining to the ledgers when pulling data from it?

**Review by nataliazm99 — APPROVED** [review]

**Pull request merged**